browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,125 @@
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/browser_window.h"
+#include "chrome/browser/ui/simple_message_box.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/grit/chromium_strings.h"
+#include "components/country_codes/country_codes.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/prefs/pref_service.h"
+#include "net/dns/public/dns_over_https_config.h"
+#include "net/dns/public/doh_provider_entry.h"
+#include "net/dns/public/secure_dns_mode.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace secure_dns = chrome_browser_net::secure_dns;
+
+namespace brave_vpn {
+
+namespace {
+const char kBraveVpnDnsProvider[] = "Cloudflare";
+
+std::string GetFilteredProvidersForCountry() {
+  namespace secure_dns = chrome_browser_net::secure_dns;
+  // Use default hardcoded servers for current country.
+  auto providers = secure_dns::ProvidersForCountry(
+      secure_dns::SelectEnabledProviders(net::DohProviderEntry::GetList()),
+      country_codes::GetCurrentCountryID());
+  for (const net::DohProviderEntry* entry : net::DohProviderEntry::GetList()) {
+    if (entry->provider != kBraveVpnDnsProvider)
+      continue;
+    net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
+    return doh_config.ToString();
+  }
+  NOTREACHED() << "Should not be reached as we expect " << kBraveVpnDnsProvider
+               << " is available in the default list.";
+  return std::string("1.1.1.1");
+}
+std::string GetDoHServers(SecureDnsConfig* dns_config) {
+  return dns_config && !dns_config->doh_servers().servers().empty()
+             ? dns_config->doh_servers().ToString()
+             : GetFilteredProvidersForCountry();
+}
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(PrefService* local_state)
+    : local_state_(local_state) {
+  pref_change_registrar_.Init(local_state);
+  pref_change_registrar_.Add(
+      prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() = default;
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
+#if !defined(OFFICIAL_BUILD)
+  if (allow_changes_for_testing_.has_value())
+    return allow_changes_for_testing_.value();
+#endif
+  auto* browser = chrome::FindLastActive();
+  return (chrome::ShowQuestionMessageBoxSync(
+              browser ? browser->window()->GetNativeWindow()
+                      : gfx::kNullNativeWindow,
+              l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+              l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_CHANGE_ALERT)) ==
+          chrome::MESSAGE_BOX_RESULT_YES);
+}
+
+void BraveVpnDnsObserverService::OnDNSPrefChanged() {
+  if (!ignore_prefs_change_)
```

Comment:
I think this should be `if (ignore_prefs_change_)` ? :) `!`

Reply:
when we enabled VPN we should ignore all external changes, so it is ok, if we do not ignore external changes we skip it

Reply:
I think logic seems fine. but it does ignore prefs change when `not ignore_prefs_change_`.

Reply:
ok, changed, thanks